README.md
# AQUATONE

Aquatone is a tool for visual inspection of websites across a large amount of hosts and is convenient for quickly gaining an overview of HTTP-based attack surface.

## Installation

1. Install [Google Chrome](https://www.google.com/chrome/) or [Chromium](https://www.chromium.org/getting-involved/download-chromium) browser -- **Note:** Google Chrome is currently giving unreliable results when running in *headless* mode, so it is recommended to install Chromium for the best results.
2. Download the [latest release](https://github.com/michenriksen/aquatone/releases/latest) of Aquatone for your operating system.
3. Uncompress the zip file and move the `aquatone` binary to your desired location. You probably want to move it to a location in your `$PATH` for easier use.

### Compiling the source code

If you for some reason don't trust the pre-compiled binaries, you can also compile the code yourself. **You are on your own if you want to do this. I do not support compiling problems. Good luck with it!**

## Usage

### Command-line options:

```
  -chrome-path string
    	Full path to the Chrome/Chromium executable to use. By default, aquatone will search for Chrome or Chromium
  -debug
    	Print debugging information
  -http-timeout int
    	Timeout in miliseconds for HTTP requests (default 3000)
  -nmap
    	Parse input as Nmap/Masscan XML
  -out string
    	Directory to write files to (default ".")
  -ports string
    	Ports to scan on hosts. Supported list aliases: small, medium, large, xlarge (default "80,443,8000,8080,8443")
  -proxy string
    	Proxy to use for HTTP requests
  -resolution string
    	screenshot resolution (default "1440,900")
  -save-body
    	Save response bodies to files (default true)
  -scan-timeout int
    	Timeout in miliseconds for port scans (default 100)
  -screenshot-timeout int
    	Timeout in miliseconds for screenshots (default 30000)
  -session string
    	Load Aquatone session file and generate HTML report
  -silent
    	Suppress all output except for errors
  -template-path string
    	Path to HTML template to use for report
  -threads int
    	Number of concurrent threads (default number of logical CPUs)
  -version
    	Print current Aquatone version
```

### Giving Aquatone data

Aquatone is designed to be as easy to use as possible and to integrate with your existing toolset with no or minimal glue. Aquatone is started by piping output of a command into the tool. It doesn't really care how the piped data looks as URLs, domains, and IP addresses will be extracted with regular expression pattern matching. This means that you can pretty much give it output of any tool you use for host discovery.

IPs, hostnames and domain names in the data will undergo scanning for ports that are typically used for web services and transformed to URLs with correct scheme.  If the data contains URLs, they are assumed to be alive and do not undergo port scanning.

**Example:**

    $ cat targets.txt | aquatone

### Output

When Aquatone is done processing the target hosts, it has created a bunch of files and folders in the current directory:

 - **aquatone_report.html**: An HTML report to open in a browser that displays all the collected screenshots and response headers clustered by similarity.
 - **aquatone_urls.txt**: A file containing all responsive URLs. Useful for feeding into other tools.
 - **aquatone_session.json**: A file containing statistics and page data. Useful for automation.
 - **headers/**: A folder with files containing raw response headers from processed targets
 - **html/**: A folder with files containing the raw response bodies from processed targets. If you are processing a large amount of hosts, and don't need this for further analysis, you can disable this with the `-save-body=false` flag to save some disk space.
 - **screenshots/**: A folder with PNG screenshots of the processed targets

The output can easily be zipped up and shared with others or archived.

#### Changing the output destination

If you don't want Aquatone to create files in the current working directory, you can specify a different location with the `-out` flag:

    $ cat hosts.txt | aquatone -out ~/aquatone/example.com

It is also possible to set a permanent default output destination by defining an environment variable:

    export AQUATONE_OUT_PATH="~/aquatone"


### Specifying ports to scan

Be default, Aquatone will scan target hosts with a small list of commonly used HTTP ports: 80, 443, 8000, 8080 and 8443. You can change this to your own list of ports with the `-ports` flag:

    $ cat hosts.txt | aquatone -ports 80,443,3000,3001

Aquatone also supports aliases of built-in port lists to make it easier for you:

 - **small**: 80, 443
 - **medium**: 80, 443, 8000, 8080, 8443 (same as default)
 - **large**: 80, 81, 443, 591, 2082, 2087, 2095, 2096, 3000, 8000, 8001, 8008, 8080, 8083, 8443, 8834, 8888
 - **xlarge**: 80, 81, 300, 443, 591, 593, 832, 981, 1010, 1311, 2082, 2087, 2095, 2096, 2480, 3000, 3128, 3333, 4243, 4567, 4711, 4712, 4993, 5000, 5104, 5108, 5800, 6543, 7000, 7396, 7474, 8000, 8001, 8008, 8014, 8042, 8069, 8080, 8081, 8088, 8090, 8091, 8118, 8123, 8172, 8222, 8243, 8280, 8281, 8333, 8443, 8500, 8834, 8880, 8888, 8983, 9000, 9043, 9060, 9080, 9090, 9091, 9200, 9443, 9800, 9981, 12443, 16080, 18091, 18092, 20720, 28017

**Example:**

    $ cat hosts.txt | aquatone -ports large


### Usage examples

Aquatone is designed to play nicely with all kinds of tools. Here's some examples:

#### Amass DNS enumeration

[Amass](https://github.com/OWASP/Amass) is currently my preferred tool for enumerating DNS. It uses a bunch of OSINT sources as well as active brute-forcing and clever permutations to quickly identify hundreds, if not thousands, of subdomains on a  domain:

```bash
$ amass -active -brute -o hosts.txt -d yahoo.com
alerts.yahoo.com
ads.yahoo.com
am.yahoo.com
- - - SNIP - - -
prd-vipui-01.infra.corp.gq1.yahoo.com
cp103.mail.ir2.yahoo.com
prd-vipui-01.infra.corp.bf1.yahoo.com
$ cat hosts.txt | aquatone
```

There are plenty of other DNS enumeration tools out there and Aquatone should work just as well with any other tool:

- [Sublist3r](https://github.com/aboul3la/Sublist3r)
- [Subfinder](https://github.com/subfinder/subfinder)
- [Knock](https://github.com/guelfoweb/knock)
- [Fierce](https://www.aldeid.com/wiki/Fierce)
- [Gobuster](https://github.com/OJ/gobuster)

#### Nmap or Masscan

Aquatone can make a report on hosts scanned with the [Nmap](https://nmap.org/) or [Masscan](https://github.com/robertdavidgraham/masscan) portscanner. Simply feed Aquatone the XML output and give it the `-nmap` flag to tell it to parse the input as Nmap/Masscan XML:

    $ cat scan.xml | aquatone -nmap

### Credits

- Thanks to [EdOverflow](https://twitter.com/EdOverflow) for the [can-i-take-over-xyz](https://github.com/EdOverflow/can-i-take-over-xyz/) project which Aquatone's domain takeover capability is based on.
- Thanks to [Elbert Alias](https://github.com/AliasIO) for the [Wappalyzer](https://github.com/AliasIO/Wappalyzer) project's technology fingerprints which Aquatone's technology fingerprinting capability is based on.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.7.0]

### Added
- New `session:start` and `session:end` events have been introduced in the event bus to allow agents to perform bootstrap
and cleanup tasks if needed
- A temporary user directory is now created for the Chrome/Chromium process and additional command line flags have been
added to increase compartmentalization

### Changed
- Production versions of Vue.js and Vue Router are now used in the HTML report for increased performance
- List of user agents have been updated to current list of most common user agents

## [1.7.0-beta.2]

### Fixed
- The pagination logic in the new HTML report would skip the page or cluster at index 0 as the `v-for` function on an integer value in Vue.js starts from 1 and not 0

## [1.7.0-beta]

### Added
- Session data will now be written to output directory as `aquatone_session.json`
- New `url_hostname_resolver` agent that resolves page's hostnames to IP addresses
- New `url_page_title_extractor` that extracts HTML page titles from responsive pages
- New command line flag `-template-path` to specify a custom template to use for the HTML report
- New command line flag `-session` to load a previous Aquatone session file and generate a report on its data
- Aquatone is now compiled for ARM64 in `build.sh`

### Changed
- Bigger refactoring of session and pages
- **New [Vue.js](https://vuejs.org/) powered HTML report with lots of new cool stuff:**
   - New look and feel
   - Pages can now be viewed in different modes:
      - **By Similarity**: Pages are displayed in clusters by their HTML structure similarity
      - **By Hostname:** Pages are displayed in clusters by their hostname
      - **Single Pages:** Pages are shown one-by-one with bigger screenshots and response headers (oldschool Aquatone style)
   - **[Vis.js](http://visjs.org/) powered network graph view** to see relations between pages, IP addresses and technologies
   - Page clusters are now rendered in a paginated carousel view instead of horizontally scrollable lanes
   - Clusters and pages are paginated to improve performance on large reports
   - Page titles are now shown for pages

### Removed
- `url_logger` agent (no longer needed)

## [1.6.0]

### Fixed
- The Nmap/Masscan XML report parser did not ignore closed/filtered ports. It now only works on ports with state `open`.

### Added
- Support for processing of multiple URLs on the same host by appending hash of URL path and fragment to file names
- Support for defining default output directory in `AQUATONE_OUT_PATH` environment variable

## [1.5.0]

### Added
- Automatic SSL/TLS detection on non-standard ports
- URL Screenshotter agent now takes extra steps to ensure that the browser process is killed after use
- Version flag to output current version (woah!!!)

### Changed
- Packages and other dependencies have been updated to latest versions
- User-Agent list has been updated to current most common agents
- Wappalyzer technology fingerprints have been updated

## [1.4.3]

### Fixed
- The Sub Resource Integrity attribute on the external CSS resource in the HTML report caused it to not load as the file had changed. Removed SRI on external CSS resource

## [1.4.2]

### Added
- Responsive URLs are now written to `aquatone_urls.txt`. Thanks [eur0pa](https://github.com/eur0pa)!
- A warning is printed when older versions of Chromium is detected which has known problems with screenshotting HTTPS URLs

### Fixed
- Aquatone had trouble processing a single or very few targets. A small delay has been added to give agents time to emit all their events

## [1.4.1]

### Changed
- List of User-Agents have been updated with most recent list of common User-Agents

### Fixed
- Random User-Agent and other spoofing request headers were not set correctly when requesting URLs. Thanks to [eur0pa](https://github.com/eur0pa) for pointing it out!

## [1.4.0]

### Added
- Passive fingerprinting of web technology in use on websites with Wappalyzer fingerprints
- Detection of domain takeover vulnerabilities across 20 different services

## [1.3.2]

Complete rewrite and simplification of Aquatone. Now written in Go and focused on reporting and screenshotting.

### Added
- Extraction of hosts, IPs and URLs from arbitrary data piped to Aquatone
- Parsing of Nmap/Masscan XML files
- Clustering of websites with similar structure in HTML report

### Removed
- Domain discovery (`aquatone-discover`)
- Domain takeover discovery (`aquatone-takeover`)

[Unreleased]: https://github.com/michenriksen/aquatone/compare/v1.7.0...HEAD
[1.7.0]: https://github.com/michenriksen/aquatone/compare/v1.7.0-beta.2...v1.7.0
[1.7.0-beta.2]: https://github.com/michenriksen/aquatone/compare/v1.7.0-beta...v1.7.0-beta.2
[1.7.0-beta]: https://github.com/michenriksen/aquatone/compare/v1.6.0...v1.7.0-beta
[1.6.0]: https://github.com/michenriksen/aquatone/compare/v1.5.0...v1.6.0
[1.5.0]: https://github.com/michenriksen/aquatone/compare/v1.4.3...v1.5.0
[1.4.3]: https://github.com/michenriksen/aquatone/compare/v1.4.2...v1.4.3
[1.4.2]: https://github.com/michenriksen/aquatone/compare/v1.4.1...v1.4.2
[1.4.1]: https://github.com/michenriksen/aquatone/compare/v1.4.0...v1.4.1
[1.4.0]: https://github.com/michenriksen/aquatone/compare/v1.3.2...v1.4.0
[1.3.2]: https://github.com/michenriksen/aquatone/compare/v1.3.2


core/urls.go
package core

import (
	"fmt"
)

var (
	securePorts = []int{443, 832, 981, 1010, 1311, 2083, 2087, 2095, 2096, 4712,
		7000, 8172, 8243, 8333, 8443, 8834, 9443, 12443, 18091, 18092}
)

func HostAndPortToURL(host string, port int, protocol string) string {
	var url string
	if protocol != "" {
		url = fmt.Sprintf("%s://%s", protocol, host)
	} else if isSecurePort(port) {
		url = fmt.Sprintf("https://%s", host)
	} else {
		url = fmt.Sprintf("http://%s", host)
	}
	if isStandardPort(port) {
		url = fmt.Sprintf("%s/", url)
	} else {
		url = fmt.Sprintf("%s:%d/", url, port)
	}
	return url
}

func isSecurePort(port int) bool {
	for _, p := range securePorts {
		if p == port {
			return true
		}
	}
	return false
}

func isStandardPort(port int) bool {
	if port == 80 || port == 443 {
		return true
	}
	return false
}
